packages/wdio-browserstack-service/src/testOps/listener.ts

```diff
@@ -0,0 +1,227 @@
+        try {
+            this.testStartedStats.triggered()
+            this.sendBatchEvents(this.getEventForHook('TestRunStarted', testData))
+        } catch (e) {
+            this.testStartedStats.failed()
+            throw e
+        }
+    }
+
+    public testFinished(testData: TestData): void {
+        try {
+            this.testFinishedStats.triggered(testData.result)
+            this.sendBatchEvents(this.getEventForHook('TestRunFinished', testData))
+        } catch (e) {
+            this.testFinishedStats.failed(testData.result)
+            throw e
+        }
+    }
+
+    public logCreated(logs: LogData[]): void {
+        this.markLogs('triggered', logs)
+        this.sendBatchEvents({
+            event_type: 'LogCreated', logs: logs
+        })
+    }
+
+    public async onScreenshot(jsonArray: ScreenshotLog[]) {
+        try {
+            this.markLogs('triggered', jsonArray)
+            // await uploadEventData([{
+            //     event_type: 'LogCreated', logs: jsonArray
+            // }], DATA_SCREENSHOT_ENDPOINT)
+            await sendScreenshots([{
+                event_type: 'LogCreated', logs: jsonArray
+            }])
+            this.markLogs('success', jsonArray)
+        } catch (e) {
+            this.markLogs('failed', jsonArray)
+            throw e
+        }
+    }
+
+    public cbtSessionCreated(data: any): void { // TODO: jsonArray any type
+        this.cbtSessionStats.triggered()
+        this.sendBatchEvents(data)
+    }
+
+    private saveData(data: any) {
+        // TODO: Remove after debugging
+        BStackLogger.debug(`data from worker is ${util.inspect(data, { depth: 6 })}`)
+
+        const logFolderPath = path.join(process.cwd(), 'logs', 'worker_data')
+        const filePath = path.join(logFolderPath, 'worker-data-' + process.pid + '.log')
+
+        if (!fs.existsSync(logFolderPath)) {
+            fs.mkdirSync(logFolderPath, { recursive: true })
+        }
+        fs.writeFileSync(filePath, JSON.stringify(data))
+    }
+
+    private markLogs(status: string, data: LogData[]): void {
+        try {
+            for (const _log of data) {
+                const kind = _log.kind
+                this.logEvents.mark(status, LOG_KIND_USAGE_MAP[kind] || kind)
+            }
+        } catch (e) {
+            BStackLogger.debug('Exception in marking logs status ' + e)
+            throw e
+        }
+    }
+
+    private getResult(jsonObject: UploadType, kind: string): string | undefined {
+        const runStr = kind === 'test' ? 'test_run' : 'hook_run'
+        const runData = jsonObject[runStr]
+        return runData?.result
+    }
+
+    private sendBatchEvents(jsonObject: UploadType): void {
+        if (!this.requestBatcher) {
+            this.requestBatcher = RequestQueueHandler.getInstance(async (data: UploadType[]) => {
+                BStackLogger.debug('callback: called with events ' + data.map(event => event.event_type))
```

Comment:
Can we store it in a variable for `data.map(event => event.event_type)` , can see it getting used multiple times

Reply:
just added this for debugging, not needed. removed